### PR TITLE
copr: support Bit column push down

### PR DESCRIPTION
**Pull request by xiejiandong**
## TiKV coprocessor: support Bit column push down
Issue: https://github.com/tikv/tikv/issues/11893
TiDB old version not support Bit column push down;  
One reason is TiKV not support,  another reason is TiDB set it unsupport. 

### This PR have Some files and functions update: 
1. components/tidb_query_aggr/src/util.rs
    rewrite_exp_for_sum_avg() 
2. components/tidb_query_datatype/src/def/eval_type.rs  
    TryFrom()
3. components/tidb_query_datatype/src/codec/chunk/column.rs
    from_raw_datums()
    from_vector_value()    
    +append_bit_datum()
4、components/tidb_query_expr/src/impl_cast.rs
     cast_bit_as_string()

TiDB's corresponding PR: pingcap/tidb#31884

### Tests
- Unit test
- [Integration test](https://github.com/tikv/copr-test)
- Manual test：[test reposrt](https://docs.google.com/document/d/1jPOzbBZlxkdusJPQfh38IW7KfDv08YERKqpMAfG7oEM/edit?usp=sharing)

### Side effects
- No side effects.
- Be backward compatibility
- Need TiDB upgrade together to use the feature.

### Release note
Can realease.

### CODE REVIEW

File: components/tidb_query_aggr/src/util.rs [hunk 1]
```diff
@@ -39,10 +39,20 @@ pub fn rewrite_exp_for_sum_avg(schema: &[FieldType], exp: &mut RpnExpression) ->
...
             return Ok(());
         }
-        EvalType::Int => FieldTypeBuilder::new()
-            .tp(FieldTypeTp::NewDecimal)
-            .flen(tidb_query_datatype::MAX_DECIMAL_WIDTH)
-            .build(),
+        EvalType::Int => {
+            if ret_field_type.tp() == FieldTypeTp::Bit {
+                FieldTypeBuilder::new()
+                    .tp(FieldTypeTp::Double)
```
Comment (skyzh) [inline components/tidb_query_aggr/src/util.rs:45]: Why is `Bit`'s field type `Double` in this part? Is this specified in TiDB?

File: components/tidb_query_datatype/src/codec/chunk/column.rs [hunk 2]
```diff
@@ -73,7 +73,15 @@ impl Column {
         let eval_type = box_try!(EvalType::try_from(field_type.as_accessor().tp()));
         match eval_type {
             EvalType::Int => {
-                if field_type.is_unsigned() {
+                if field_type.as_accessor().tp() == FieldTypeTp::Bit {
+                    if field_type.as_accessor().flen() > 64 {
+                        unimplemented!()
+                    }
+                    let start_idx: usize = ((64 - field_type.as_accessor().flen()) / 8) as usize;
```
Comment (skyzh) [inline components/tidb_query_datatype/src/codec/chunk/column.rs:80]: Is there any document on `Bit` vector (and scalar)'s encoding? These operations that doesn't look straightforward to me, and I guess it's a kind of hack to make this work.

**Review by skyzh — COMMENTED** [review]
Rest LGTM.

What's TiDB's corresponding PR to this TiKV PR?

And the same time, would you please make the test report accessible by URL, so that community members can look into that?

File: components/tidb_query_datatype/src/codec/chunk/column.rs (hunk 2 above)
Comment (xiejiandong) [reply components/tidb_query_datatype/src/codec/chunk/column.rs:80]: Bit is defined bits range from 1 to 64:  https://dev.mysql.com/doc/refman/8.0/en/bit-type.html
For higher version of mysql, Bit will be more bytes, but TiDB has't support yet，and TiDB defined Bit storage len be 8 bytes（tidb/parser/types/field_type.go +352）now，hard code； So，for future TiDB that Bit support more bytes，we should fix TiKV codes here；
Maybe I should has some comments here.

File: components/tidb_query_aggr/src/util.rs (hunk 1 above)
Comment (xiejiandong) [reply components/tidb_query_aggr/src/util.rs:45]: Codes tidb\expression\aggregation\base_func.go  define typeInfer4Sum()、typeInfer4Avg(), For Bit RetType is Double.
TiDB's corresponding PR:  https://github.com/pingcap/tidb/pull/31884

**Review by skyzh — APPROVED** [review]
LGTM from my side -- you'll need another reviewer for this PR.

At the same time, you should add this case to integration-copr-test. This check is expected to run for every PR, but it seems that the EE team has removed it. Anyway, you can merge the PRs in TiDB and TiKV first, and then add push-down integration test (instead of doing manual test every time).

See https://github.com/tikv/copr-test for copr-test examples and information.

Good work!

**Comment by xiejiandong** [comment]
status/can-merge

File: components/tidb_query_datatype/src/def/eval_type.rs [hunk 3]
```diff
@@ -75,7 +76,7 @@ impl std::convert::TryFrom<crate::FieldTypeTp> for EvalType {
             | crate::FieldTypeTp::Null => EvalType::Bytes,
             crate::FieldTypeTp::Enum => EvalType::Enum,
             _ => {
-                // In TiDB, Bit's eval type is Int, but it is not yet supported in TiKV.
+                // In TiDB, Bit's eval type is Int
```
Comment (zhongzc) [inline components/tidb_query_datatype/src/def/eval_type.rs:79]:
```suggestion
```

File: components/tidb_query_datatype/src/codec/chunk/column.rs (hunk 2 above)
Comment (zhongzc) [reply components/tidb_query_datatype/src/codec/chunk/column.rs:82]: Comments are needed for this line of code.

File: components/tidb_query_datatype/src/codec/chunk/column.rs [hunk 4]
```diff
@@ -523,6 +552,27 @@ impl Column {
...
+    pub fn append_bit_datum(&mut self, src_datum: &[u8], start_idx: &usize) -> Result<()> {
+        if src_datum.is_empty() {
+            return Err(Error::InvalidDataType(
+                "Failed to decode datum flag".to_owned(),
+            ));
+        }
+        let flag = src_datum[0];
+        match flag {
+            datum::NIL_FLAG => self.append_null(),
+            datum::UINT_FLAG => self.append_bytes(&src_datum[(1 + start_idx)..])?,
```
Comment (zhongzc) [inline components/tidb_query_datatype/src/codec/chunk/column.rs:565]: How `1 +` works? Need comments I think.

Comment (xiejiandong) [reply components/tidb_query_datatype/src/codec/chunk/column.rs:565]: It skip [0] flag; ok;

File: components/tidb_query_datatype/src/codec/chunk/column.rs (hunk 2 above)
Comment (xiejiandong) [reply components/tidb_query_datatype/src/codec/chunk/column.rs:84]: ok, has added it.

**Pull request closed without merging**